libsoftwarecontainer/src/gateway/cgroups/cgroupsparser.cpp

```diff
@@ -27,25 +28,94 @@ CGroupsParser::CGroupsParser()
+{
+    char *end;
+    long long limit_in_bytes = std::strtoll(settingValue.c_str(), &end, 10);
+    if (errno == ERANGE){
+        log_error() << settingValue << " is out of range";
+        throw LimitRangeError();
+    }
+
+    if (limit_in_bytes * multiply > std::numeric_limits<long long>::max()) {
+        log_error() << settingValue << " * " << std::to_string(multiply) <<" is out of range";
+        throw LimitRangeError();
+    }
+
+    limit_in_bytes *= multiply;
+    return std::to_string(limit_in_bytes);
+}
+
+std::string CGroupsParser::suffixCorrection(const std::string settingValue)
+{
+    auto suffix = settingValue.back();
+    std::string removeUnit = settingValue;
+    removeUnit.pop_back();
+
+    if (std::isdigit(suffix)) {
+        //the value is digit already no need to do anything
+        return settingValue;
+    } else if ('K' == suffix || 'k' == suffix) {
+        //the value is Kilobyte unit convert it to byte
+        int KB = 1024;
+
+        try {
+            return byteRegulation(removeUnit, KB);
+        } catch (LimitRangeError &error) {
+            throw LimitRangeError();
+        }
+    } else if ('M' == suffix || 'm' == suffix) {
+        //the value is Megabyte unit convert it to byte
+        int MB = 1024 * 1024;
+
+        try {
+            return byteRegulation(removeUnit, MB);
+        } catch (LimitRangeError &error) {
+            throw LimitRangeError();
+        }
+    } else if ('G' == suffix || 'g' == suffix) {
+        //the value is Gigabyte unit convert it to byte
+        int GB = 1024 * 1024 * 1024;
+
+        try {
+            return byteRegulation(removeUnit, GB);
+        } catch (LimitRangeError &error) {
+            throw LimitRangeError();
+        }
+    }
+    log_error() << "Bad suffix on setting value " << settingValue;
+    throw BadSuffixError("Bad suffix on setting value " + settingValue);
+}
+
+void CGroupsParser::parseCGroupsGatewayConfiguration(const json_t *element)
 {
     std::string settingKey;
     std::string settingValue;
 
     if (!JSONParser::read(element, "setting", settingKey)) {
         log_error() << "Key \"setting\" either not a string or not in json configuration";
-        return ReturnCode::FAILURE;
+        throw JSonError();
     }
 
     if (!JSONParser::read(element, "value", settingValue)) {
         log_error() << "Key \"value\" either not a string or not in json configuration";
-        return ReturnCode::FAILURE;
+        throw JSonError();
     }
 
-    if ("memory.limit_in_bytes" == settingKey) {
+    if (("memory.limit_in_bytes" == settingKey) || ("memory.memsw.limit_in_bytes" == settingKey)) {
+        try {
+            settingValue = suffixCorrection(settingValue);
+        } catch (LimitRangeError &error) {
+            log_error() << settingValue << " is not appropriate value for " << settingKey;
+            throw LimitRangeError();
+        } catch (BadSuffixError  &error) {
+            throw BadSuffixError();
```

Comment:
Use a message here (or don't catch at all, since you're just re-throwing the same type of exception)